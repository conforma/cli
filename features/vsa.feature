Feature: VSA generation and storage
  The ec command line should generate and store Verification Summary Attestation (VSA)

  Background:
    Given a stub cluster running
    Given stub rekord running
    Given stub registry running
    Given stub git daemon running
    Given stub tuf running

  Scenario: VSA generation with local storage backend
    Given a key pair named "vsa-test"
    Given an image named "acceptance/vsa-test-image"
    Given a valid image signature of "acceptance/vsa-test-image" image signed by the "vsa-test" key
    Given a valid Rekor entry for image signature of "acceptance/vsa-test-image"
    Given a valid attestation of "acceptance/vsa-test-image" signed by the "vsa-test" key
    Given a valid Rekor entry for attestation of "acceptance/vsa-test-image"
    Given a git repository named "vsa-policy" with
      | main.rego | examples/happy_day.rego |
    Given policy configuration named "vsa-ec-policy" with specification
    """
    {
      "sources": [
        {
          "policy": [
            "git::https://${GITHOST}/git/vsa-policy.git"
          ]
        }
      ]
    }
    """
    When ec command is run with "validate image --image ${REGISTRY}/acceptance/vsa-test-image --policy acceptance/vsa-ec-policy --public-key ${vsa-test_PUBLIC_KEY} --rekor-url ${REKOR} --vsa --vsa-signing-key ${vsa-test_PRIVATE_KEY} --vsa-upload local@${TMPDIR}/vsa-output --output json"
    Then the exit status should be 0
    Then the output should match the snapshot
    And VSA envelope files should exist in "${TMPDIR}/vsa-output"

  Scenario: VSA generation with Rekor storage backend
    Given a key pair named "vsa-rekor"
    Given an image named "acceptance/vsa-rekor-image"
    Given a valid image signature of "acceptance/vsa-rekor-image" image signed by the "vsa-rekor" key
    Given a valid Rekor entry for image signature of "acceptance/vsa-rekor-image"
    Given a valid attestation of "acceptance/vsa-rekor-image" signed by the "vsa-rekor" key
    Given a valid Rekor entry for attestation of "acceptance/vsa-rekor-image"
    Given a git repository named "vsa-rekor-policy" with
      | main.rego | examples/happy_day.rego |
    Given policy configuration named "vsa-rekor-ec-policy" with specification
    """
    {
      "sources": [
        {
          "policy": [
            "git::https://${GITHOST}/git/vsa-rekor-policy.git"
          ]
        }
      ]
    }
    """
    Given VSA upload to Rekor should be expected
    When ec command is run with "validate image --image ${REGISTRY}/acceptance/vsa-rekor-image --policy acceptance/vsa-rekor-ec-policy --public-key ${vsa-rekor_PUBLIC_KEY} --rekor-url ${REKOR} --vsa --vsa-signing-key ${vsa-rekor_PRIVATE_KEY} --vsa-upload rekor@${REKOR} --output json"
    Then the exit status should be 0
    Then the output should match the snapshot
    And VSA should be uploaded to Rekor successfully

  Scenario: VSA generation with multiple storage backends
    Given a key pair named "vsa-multi"
    Given an image named "acceptance/vsa-multi-image"
    Given a valid image signature of "acceptance/vsa-multi-image" image signed by the "vsa-multi" key
    Given a valid Rekor entry for image signature of "acceptance/vsa-multi-image"
    Given a valid attestation of "acceptance/vsa-multi-image" signed by the "vsa-multi" key
    Given a valid Rekor entry for attestation of "acceptance/vsa-multi-image"
    Given a git repository named "vsa-multi-policy" with
      | main.rego | examples/happy_day.rego |
    Given policy configuration named "vsa-multi-ec-policy" with specification
    """
    {
      "sources": [
        {
          "policy": [
            "git::https://${GITHOST}/git/vsa-multi-policy.git"
          ]
        }
      ]
    }
    """
    Given VSA upload to Rekor should be expected
    When ec command is run with "validate image --image ${REGISTRY}/acceptance/vsa-multi-image --policy acceptance/vsa-multi-ec-policy --public-key ${vsa-multi_PUBLIC_KEY} --rekor-url ${REKOR} --vsa --vsa-signing-key ${vsa-multi_PRIVATE_KEY} --vsa-upload local@${TMPDIR}/vsa-multi-output --vsa-upload rekor@${REKOR} --output json"
    Then the exit status should be 0
    Then the output should match the snapshot
    And VSA envelope files should exist in "${TMPDIR}/vsa-multi-output"
    And VSA should be uploaded to Rekor successfully

  Scenario: VSA generation with invalid storage backend configuration
    Given a key pair named "vsa-invalid"
    Given an image named "acceptance/vsa-invalid-image"
    Given a valid image signature of "acceptance/vsa-invalid-image" image signed by the "vsa-invalid" key
    Given a valid Rekor entry for image signature of "acceptance/vsa-invalid-image"
    Given a valid attestation of "acceptance/vsa-invalid-image" signed by the "vsa-invalid" key
    Given a valid Rekor entry for attestation of "acceptance/vsa-invalid-image"
    Given a git repository named "vsa-invalid-policy" with
      | main.rego | examples/happy_day.rego |
    Given policy configuration named "vsa-invalid-ec-policy" with specification
    """
    {
      "sources": [
        {
          "policy": [
            "git::https://${GITHOST}/git/vsa-invalid-policy.git"
          ]
        }
      ]
    }
    """
    When ec command is run with "validate image --image ${REGISTRY}/acceptance/vsa-invalid-image --policy acceptance/vsa-invalid-ec-policy --public-key ${vsa-invalid_PUBLIC_KEY} --rekor-url ${REKOR} --vsa --vsa-signing-key ${vsa-invalid_PRIVATE_KEY} --vsa-upload invalid-backend@somewhere --output json"
    Then the exit status should be 0
    Then the output should match the snapshot
