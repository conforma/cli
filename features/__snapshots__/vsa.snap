
[VSA generation with invalid storage backend configuration:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-invalid-image@sha256:${REGISTRY_acceptance/vsa-invalid-image:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-invalid-image}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-invalid-image}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-invalid_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-invalid-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-invalid_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA generation with invalid storage backend configuration:stderr - 1]
time="${TIMESTAMP}" level=warning msg="invalid storage config 'invalid-backend@somewhere': unsupported backend 'invalid-backend'. Supported backends: rekor, local"
time="${TIMESTAMP}" level=warning msg="invalid storage config 'invalid-backend@somewhere': unsupported backend 'invalid-backend'. Supported backends: rekor, local"

---

[VSA generation with local storage backend:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-test-image@sha256:${REGISTRY_acceptance/vsa-test-image:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-test-image}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-test-image}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-test_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-test_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA generation with local storage backend:stderr - 1]

---

[VSA generation with Rekor storage backend:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-rekor-image@sha256:${REGISTRY_acceptance/vsa-rekor-image:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-rekor-image}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-rekor-image}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-rekor_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-rekor-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-rekor_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA generation with Rekor storage backend:stderr - 1]

---

[VSA generation with multiple storage backends:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-multi-image@sha256:${REGISTRY_acceptance/vsa-multi-image:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-multi-image}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-multi-image}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-multi_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-multi-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-multi_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA generation with multiple storage backends:stderr - 1]

---

[VSA generation with multiple images:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-multiple-image-2@sha256:${REGISTRY_acceptance/vsa-multiple-image-2:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-multiple-image-2}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-multiple-image-2}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-multiple_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-multiple-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-multiple_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA generation with multiple images:stderr - 1]

---

[VSA expiration flag functionality:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-expiration-image@sha256:${REGISTRY_acceptance/vsa-expiration-image:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-expiration-image}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-expiration-image}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-expiration_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-expiration-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-expiration_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA expiration flag functionality:stderr - 1]
time="${TIMESTAMP}" level=warning msg="Failed to check for existing VSA for image ${REGISTRY}/acceptance/vsa-expiration-image@sha256:${REGISTRY_acceptance/vsa-expiration-image:latest_DIGEST}: failed to retrieve VSA envelope: no entries found in Rekor for image digest: sha256:${REGISTRY_acceptance/vsa-expiration-image:latest_DIGEST}"

---

[VSA expiration with existing valid VSA:stdout - 1]
{
  "success": true,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "${REGISTRY}/acceptance/vsa-existing-image@sha256:${REGISTRY_acceptance/vsa-existing-image:latest_DIGEST}",
      "source": {},
      "success": true,
      "signatures": [
        {
          "keyid": "",
          "sig": "${IMAGE_SIGNATURE_acceptance/vsa-existing-image}"
        }
      ],
      "attestations": [
        {
          "type": "https://in-toto.io/Statement/v0.1",
          "predicateType": "https://slsa.dev/provenance/v0.2",
          "predicateBuildType": "https://tekton.dev/attestations/chains/pipelinerun@v2",
          "signatures": [
            {
              "keyid": "",
              "sig": "${ATTESTATION_SIGNATURE_acceptance/vsa-existing-image}"
            }
          ]
        }
      ]
    }
  ],
  "key": "${vsa-existing_PUBLIC_KEY_JSON}",
  "policy": {
    "sources": [
      {
        "policy": [
          "git::${GITHOST}/git/vsa-existing-policy.git?ref=${LATEST_COMMIT}"
        ]
      }
    ],
    "rekorUrl": "${REKOR}",
    "publicKey": "${vsa-existing_PUBLIC_KEY}"
  },
  "ec-version": "${EC_VERSION}",
  "effective-time": "${TIMESTAMP}"
}
---

[VSA expiration with existing valid VSA:stderr - 1]

---
